### PR TITLE
POS: Receipt Printer

### PR DESCRIPTION
**Pull request by Talej**
# Description

Receipt printing functionality for android. Ended up using react-native-print package with a HTML template for receipt layout. The layout should pretty closely match the details shown on the 'Order' view for a paid order.

A print receipt button is shown on the payment completed view and also on the order view for paid orders if:

1. It's an android device
2. invoice has been paid
3. receipt printing isn't disabled in settings

react-native-print doesn't have any way of detecting if there is a printer ahead of calling the print dialog so to handle this I added the disable printer option under the POS settings.

I thought it could possibly also be useful to allow the receipt template HTML to be configuration in the POS settings - although given it is HTML this could be somewhat painful the edit on a device so wanted your input on this...

This pull request is categorized as a:

- [X] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Configuration change
- [ ] Locales update
- [ ] Quality assurance
- [ ] Other

## Checklist
- [ ] I’ve run `yarn run tsc` and made sure my code compiles correctly
- [X] I’ve run `yarn run lint` and made sure my code didn’t contain any problematic patterns
- [X] I’ve run `yarn run prettier` and made sure my code is formatted correctly
- [X] I’ve run `yarn run test` and made sure all of the tests pass

## Testing

If you modified or added a utility file, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [X] N/A

I have tested this PR on the following platforms (please specify OS version and phone model/VM):

- [X] Android
- [ ] iOS

I have tested this PR with the following types of nodes (please specify node version and API version where appropriate):

- [X] Embedded LND
- [ ] LND (REST)
- [ ] LND (Lightning Node Connect)
- [ ] Core Lightning (c-lightning-REST)
- [ ] LndHub
- [ ] [DEPRECATED] Core Lightning (Spark)
- [ ] [DEPRECATED] Eclair

### Locales
- [X] I’ve added new locale text that requires translations
- [ ] I’m aware that new translations should be made on the ZEUS [Transfix page](https://app.transifex.com/ZeusLN/zeus/) and not directly to this repo

### Third Party Dependencies and Packages

- [X] Contributors will need to run `yarn` after this PR is merged in
- [X] 3rd party dependencies have been modified:
    * verify that `package.json` and `yarn.lock` have been properly updated
    * verify that dependencies are installed for both iOS and Android platforms

### Other:

- [ ] Changes were made that require an update to the README
- [ ] Changes were made that require an update to onboarding

### CODE REVIEW

**Comment by Talej** [comment]
Screenshot_20240127-135755
Screenshot_20240127-135917
Screenshot_20240127-135946

**Comment by kaloudis** [comment]
I wonder if it's worth distinguishing `Print Invoice` / `Print Receipt` depending on where we are in the flow

**Comment by Talej** [comment]
> I wonder if it's worth distinguishing `Print Invoice` / `Print Receipt` depending on where we are in the flow

Do you mean have a `Print Invoice` on the order screen before it is paid and `Print Receipt` after paid?

**Comment by kaloudis** [comment]
> > I wonder if it's worth distinguishing `Print Invoice` / `Print Receipt` depending on where we are in the flow
> 
> Do you mean have a `Print Invoice` on the order screen before it is paid and `Print Receipt` after paid?

Precisely, and perhaps we add the payment pre-image on the receipt to distinguish them and act as a bit of payment proof

File: package.json [hunk 1]
```diff
@@ -111,6 +111,7 @@
     "react-native-notifications": "5.1.0",
     "react-native-os": "aprock/react-native-os#5/head",
     "react-native-permissions": "3.8.0",
+    "react-native-print": "^0.11.0",
```
Comment (kaloudis) [inline package.json:114]: let's lock in the version number here

File: views/Order.tsx [hunk 2]
```diff
@@ -312,6 +352,152 @@ export default class OrderView extends React.Component<OrderProps, OrderState> {
...
+            return UnitsStore.getFormattedAmount(amount, units);
+        };
+
+        const printReceipt = () => {
+            const title =
+                merchantName && merchantName.length
+                    ? merchantName
+                    : isPaid
+                    ? 'Tax Receipt'
+                    : 'Invoice';
```
Comment (kaloudis) [inline views/Order.tsx:393]: We need these titles localized

Comment (Talej) [reply views/Order.tsx]: Done!

File: package.json (hunk 1 above)
Comment (Talej) [reply package.json]: Done!

**Comment by kaloudis** [comment]
Nice job only showing the printer setting for Android, but I think it would be better to make it opt in rather than opt out as most merchants likely won't have a printer module.

**Comment by Talej** [comment]
> Nice job only showing the printer setting for Android, but I think it would be better to make it opt in rather than opt out as most merchants likely won't have a printer module.

Makes total sense. Flipped the setting from disable to enable.

File: locales/en.json [hunk 3]
```diff
@@ -934,9 +936,13 @@
     "pos.views.Order.totalBitcoin": "Total (Bitcoin)",
     "pos.views.Order.total": "Total",
     "pos.views.Order.paymentType": "Payment type",
+    "pos.views.Order.printReceipt": "Print Receipt",
+    "pos.views.Order.printInvoice": "Print Invoice",
     "pos.views.Settings.PointOfSale.authWarning": "Warning: no password or PIN set",
     "pos.views.Settings.PointOfSale.backendWarning": "Warning: currently only LND nodes are able to mark orders as paid",
     "pos.views.Settings.PointOfSale.currencyError": "Error: currency must be set first",
+    "pos.print.taxReceipt": "Tax Receipt",
```
Comment (kaloudis) [inline locales/en.json:944]: nit: can we just make this `Receipt` instead of `Tax Receipt`

**Comment by kaloudis** [comment]
@Talej can you provide an example of what these receipts look like when the merchant provides their store name in settings?

**Comment by Talej** [comment]
> @Talej can you provide an example of what these receipts look like when the merchant provides their store name in settings?

proof-of-storename

IMG_1147

**Review by shubhamkmr04 — COMMENTED** [review]
Screenshot_1708081156
Screenshot_1708081167

It would be better if we give a paddingBottom here.

**Review by shubhamkmr04 — COMMENTED** [review]
Screenshot_1708081031
Also can we also handle the case so the user don't get this error if the device is not connected to a printer.

**Comment by Talej** [comment]
> Screenshot_1708081031 Also can we also handle the case so the user don't get this error if the device is not connected to a printer.

This actually looks like the package isn't loading for you rather than anything to do with not connecting to a printer.

I have tried with various devices and even if there is no printer it still works without error (it gives an option to export as PDF). I have added in some extra checks to prevent the error but would be useful if you could debug a bit further to see why it wasn't loading.

**Review by shubhamkmr04 — APPROVED** [review]
tACK

**Review by kaloudis — APPROVED** [review]
LGTM. Great job, @Talej!

**Pull request merged**